src/protocols/secure_channel/CheckinMessage.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements the utility class for the Matter Checkin protocol.
+ */
+
+#include "CheckinMessage.h"
+#include <lib/core/CHIPCore.h>
+
+namespace chip {
+namespace Protocols {
+namespace SecureChannel {
+
+static const uint16_t mMaxPlainTextSize = CHIP_CHECK_IN_APP_DATA_MAX_SIZE + CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES;
+static const uint16_t mMaxWorBufferSize = mMaxPlainTextSize + sizeof(uint32_t);
+static const uint16_t minPayloadSize = CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES + sizeof(uint32_t) + CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES;
+
+CHIP_ERROR CheckinMessage::GenerateCheckingMessagePayload(Crypto::Aes128KeyHandle & key, uint32_t counter, ByteSpan appData,
+                                                          uint8_t * output, uint16_t * payloadSize)
+{
+    VerifyOrReturnError(nullptr != output, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(appData.size() <= CHIP_CHECK_IN_APP_DATA_MAX_SIZE, CHIP_ERROR_INVALID_ARGUMENT);
+
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    uint8_t * outputPtr = output;
+
+    {
+        chip::Crypto::HMAC_sha shaHandler;
+        uint8_t nonceWorkBuffer[CHIP_CRYPTO_HASH_LEN_BYTES] = { 0 };
+        ReturnErrorOnFailure(shaHandler.HMAC_SHA256(key.As<Aes128KeyByteArray>(), sizeof(Aes128KeyByteArray) * 8,
+                                                    (uint8_t *) &counter, sizeof(counter), nonceWorkBuffer,
```

Comment:
This does not match spec, as far as I can tell.  Please check carefully.  For example, spec says counter encoded little-endian into bytes...

But also, where is `sizeof(Aes128KeyByteArray) * 8` coming from?  I would expect key length to be in bytes, not bits... though clearly we need to better document this in CryptoPAL.  What I don't understand is how things passed tests with this.